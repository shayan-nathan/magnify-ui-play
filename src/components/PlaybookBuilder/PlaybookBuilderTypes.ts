import { DragEvent } from 'react'
import { Node } from 'react-flow-renderer'

import { MetadataDescription } from 'models/metadata.model'

export type shapeType = 'circle' | 'rhomb' | 'square'
export interface Action {
  name: string
  type: string
  shape: shapeType | string
  platform?: string
  iconName?: string
  payload?: SegmentBuilderData | {}
  description?: string
}
export interface Platform {
  name: string
  type: string
  actions: Action[]
}
export interface PlaybookActionsType {
  targeting: Action[]
  rules: Action[]
  actions?: Platform[]
}

export interface LooseObject {
  [key: string]: any
}
export interface BuilderAction extends Action {
  id: string
  source: string
  target: string
  data: {
    [k: string]: any
  }
}

interface SegmentBuilderProps {
  visible: boolean
  onClose: () => void
  segmentBuilderData: any
}

interface SegmentSidebarProps {
  handleDragStart: (e: DragEvent<HTMLDivElement>, item: SourceData, breadCrumbItem: any[], position: any) => void
  handleDragEnter: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
  handleDragEnd: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
  handleDragOver: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
}

interface SegmentCriteriaProps {
  data: any
  dragEvents: DragEvents
  segmentBuilderData: any
  handleDragStart: (e: DragEvent<HTMLDivElement>, item: SourceData, breadCrumbItem: any[], position: any) => void
  handleDragEnter: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
  onDragEnterGroup: (e: DragEvent<HTMLDivElement>, group?: number | null) => void
  onDragLeaveGroup: (e: DragEvent<HTMLDivElement>, group?: number | null) => void
  onDragOverGroup: (e: DragEvent<HTMLDivElement>, group?: number | null) => void
  handleDragLeave: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
  handleDragOver: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
  handleDrop: (e: DragEvent<HTMLDivElement>, item?: SourceData, position?: any) => void
  onDragEnterPlaceholder: (e: DragEvent<HTMLDivElement>, group?: number | null) => void
  onDragLeavePlaceholder: (e: DragEvent<HTMLDivElement>, group?: number | null) => void
  onDragOverPlaceholder: (e: DragEvent<HTMLDivElement>, group?: number | null) => void
}

interface breadcrumbInfo extends MetadataDescription {
  path: string[]
}

interface SourceData {
  name: string
  type?: string
  fields?: SourceData[]
}

interface Breadcrumb {
  name: string
  type: string
  path: string[]
}

interface DragItemEvent {
  name: string
  item: SourceData
}

interface Field5 {
  name: string
  type: string
  fields: any[]
}

interface Field4 {
  name: string
  type: string
  fields: Field5[]
}

interface Field3 {
  name: string
  type: string
  fields: Field4[]
}

interface Field2 {
  name: string
  type?: string
  fields?: Field3[]
}

interface Field {
  name: string
  type?: string
  fields?: Field2[]
}

interface mockDataItems {
  name: string
  type: string
  fieldType?: string
  fields: Field[]
}
interface CreateEdgeProps {
  parentNode: Node
  childNode: Node | string
  dropEdge?: BuilderAction | string
  isOnNoBranch?: boolean
}

interface DragEvents {
  isDragStarting: boolean
  dragOverGroup: number | undefined
  groupCounter: number
  placeholderCounter: number
}

interface SegmentBuilderData {
  groups: Groups[]
  relationOperator: string
}

interface Groups {
  groups: Item[]
  relationOperator: string
}

interface Item {
  platform: string
  object: string
  type: string | undefined
  field: string | undefined
  operator: string
  value: string
  isNewStatement: boolean
}
interface SelectOptions {
  label: string
  value: string
}

interface SegmentDescriptionProps {
  segmentBuilderData: Action
  visible: boolean
}

export type {
  SegmentSidebarProps,
  breadcrumbInfo,
  SourceData,
  SegmentBuilderProps,
  CreateEdgeProps,
  SegmentCriteriaProps,
  DragItemEvent,
  mockDataItems,
  DragEvents,
  SegmentBuilderData,
  Groups,
  Item,
  SelectOptions,
  SegmentDescriptionProps,
}
