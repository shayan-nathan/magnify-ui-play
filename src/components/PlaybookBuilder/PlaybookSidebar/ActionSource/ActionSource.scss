$builder-action-color: #3e4558;

.action-source {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-items: center;

  .c-targeting {
    display: flex;
    align-items: center;
    justify-content: center;
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.08);
    border: 1px solid rgba(205, 208, 216, 0.5);
    padding: 0;

    path,
    rect {
      fill: $builder-action-color;
    }

    circle {
      stroke: $builder-action-color;
    }

    g {
      path {
        stroke: $builder-action-color;
      }
    }

    &--circle {
      border-radius: 50%;
      width: 52px;
      height: 52px;
    }

    &--rhomb {
      border-radius: 6px;
      transform: rotate(45deg);
      width: 40px;
      height: 40px;
      margin-bottom: 4px;

      svg {
        transform: rotate(-45deg);
      }
    }

    &--square {
      width: 44px;
      height: 44px;
      border-radius: 6px;
    }
  }

  .action-btn--box {
    width: 60px;
    height: 60px;
    display: flex;
    align-items: center;
    justify-content: center;
  }

  .source-name {
    padding-left: 8px;
  }

  .c-disabled {
    opacity: 0.4;
  }
}
