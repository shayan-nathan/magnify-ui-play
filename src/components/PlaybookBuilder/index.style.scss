@import './PlaybookSidebar/PlaybookSidebar.scss';
@import './SegmentBuilder/SegmentBuilder.scss';
