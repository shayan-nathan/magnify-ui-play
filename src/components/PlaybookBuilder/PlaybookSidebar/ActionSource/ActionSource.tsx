import { Button } from 'antd'
import { FC, DragEvent } from 'react'
import classNames from 'classnames'

import { getNormalize } from 'services/Utils/parseString.utils'
import { Action } from 'components/PlaybookBuilder/PlaybookBuilderTypes'
import { useStore } from 'store'
import { BuilderIcon } from 'services/Utils/BuilderIcon'

interface ActionSourceProps {
  action: Action
}

export const ActionSource: FC<ActionSourceProps> = (props: ActionSourceProps) => {
  const { action } = props
  const shape = action.shape

  const { playbookStore } = useStore()

  const onDragStart = (event: DragEvent, nodeType: string) => {
    playbookStore.setDisplayEdgeDrops(true)
    playbookStore.setPlaceholderShape(shape)
    event.currentTarget.className += ' c-disabled'
    event.dataTransfer.setData('application/reactflow', nodeType)
    event.dataTransfer.setData('application/json', JSON.stringify(action))
    event.dataTransfer.effectAllowed = 'copy'
  }

  const onDragEnd = (event: DragEvent) => {
    // console.log('drag end', event)
    playbookStore.setDisplayEdgeDrops(false)
    event.currentTarget.className = event.currentTarget.className.replaceAll(' c-disabled', '')
  }

  return (
    <div
      className='action-source'
      data-testid='action-source'
      draggable
      onDragStart={(event: DragEvent) => onDragStart(event, 'segment')}
      onDragEnd={(event: DragEvent) => {
        return onDragEnd(event)
      }}>
      <div className='action-btn--box' data-testid='source-box'>
        <Button className={classNames([`c-targeting c-targeting--${shape || 'square'}`])} data-testid='source-btn'>
          <BuilderIcon name={`${getNormalize(action.type)}`} options={{ width: 22, height: 22 }} />
          {/* {!loading && <img src={`${image}`} alt={action.name} draggable={false} data-testid='source-image' />} */}
        </Button>
      </div>
      <div className='source-name' data-testid='source-name'>
        <span>{action.name}</span>
      </div>
    </div>
  )
}
