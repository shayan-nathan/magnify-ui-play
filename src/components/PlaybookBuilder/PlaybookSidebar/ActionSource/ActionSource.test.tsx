import { render, screen } from '@testing-library/react'
import { Action } from 'components/PlaybookBuilder/PlaybookBuilderTypes'

import { ActionSource } from './ActionSource'

const mockProps: Action = {
  name: 'Segments',
  type: 'segment',
  shape: 'circle',
}

describe('<ActionSource /> component', () => {
  it('renders correctly', async () => {
    render(<ActionSource action={mockProps} />)
    const component = await screen.findByTestId('action-source')
    expect(component).toMatchSnapshot()
  })

  it('has an active button that is not draggable and the action box draggable', async () => {
    render(<ActionSource action={mockProps} />)
    const button = await screen.findByTestId('source-btn')
    // check button to not be disable
    expect(button).not.toBeDisabled()
    // check button to not be draggable
    expect(button).toHaveProperty('draggable', false)
    // check button to have the right style
    const sourceBoxElement = await screen.findByTestId('action-source')
    expect(sourceBoxElement).toHaveProperty('draggable', true)
  })

  it('has a button with the shape style class', async () => {
    render(<ActionSource action={mockProps} />)
    const button = await screen.findByTestId('source-btn')
    expect(button).toHaveClass(`c-targeting--${mockProps.shape}`)
  })

  it('display the name of the action', async () => {
    render(<ActionSource action={mockProps} />)
    const element = await screen.findByTestId('source-name')
    expect(element.textContent).toEqual(mockProps.name)
  })

  it('displaythe action image', async () => {
    render(<ActionSource action={mockProps} />)
    const image = await screen.findByTestId('source-image')
    expect(image).toHaveAttribute('src', 'segment.svg')
    expect(image).toHaveAttribute('alt', mockProps.name)
  })

  it('style the component based on type', async () => {
    render(<ActionSource action={mockProps} />)
    const image = await screen.findByTestId('source-image')
    expect(image).toHaveAttribute('src', 'segment.svg')
    expect(image).toHaveAttribute('alt', mockProps.name)
  })
})
