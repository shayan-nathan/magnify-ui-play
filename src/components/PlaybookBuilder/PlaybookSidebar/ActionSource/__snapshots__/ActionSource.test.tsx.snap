// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionSource /> component renders correctly 1`] = `
<div
  class="action-source"
  data-testid="action-source"
  draggable="true"
>
  <div
    class="action-btn--box"
    data-testid="source-box"
  >
    <button
      class="ant-btn c-targeting c-targeting--circle"
      data-testid="source-btn"
      type="button"
    >
      <img
        alt="Segments"
        data-testid="source-image"
        draggable="false"
        src="segments.svg"
      />
    </button>
  </div>
  <div
    class="source-name"
    data-testid="source-name"
  >
    <span>
      Segments
    </span>
  </div>
</div>
`;
